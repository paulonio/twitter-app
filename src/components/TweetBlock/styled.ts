import { styled } from 'styled-components';

export const TweetWrapper = styled.section`
  padding: 15px 11px;
  display: flex;
  justify-content: space-between;
  border-top: solid 1px #d8d8d8;
  border-bottom: solid 1px #d8d8d8;
`;

export const AvatarWrapper = styled.div`
  width: 49px;
  height: 54px;
  margin-left: 17px;
`;

export const TweetContent = styled.div`
  width: 87%;
  max-width: 794px;
  display: flex;
  flex-direction: column;
  align-items: start;
  justify-content: space-between;
`;

export const TweetInput = styled.input.attrs({ placeholder: "What's happening" })`
  width: 100%;
  font-size: 22px;
  line-height: 29px;
  font-weight: 600;
  color: #828282;
  border: none;
  outline: none;
`;

export const TweetFooter = styled.div`
  width: 100%;
  margin-top: 38px;
  display: flex;
  align-items: center;
  justify-content: space-between;
`;

export const ImageIcon = styled.img.attrs({
  src: '/assets/icons/ImageIcon.svg',
  alt: 'Image icon',
})`
  width: 24px;
  height: 24px;
  &:hover {
    cursor: pointer;
  }
`;

export const ButtonWrapper = styled.div`
  width: 13%;
  max-width: 108px;
`;
