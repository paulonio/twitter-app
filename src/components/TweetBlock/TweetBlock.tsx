import React, { FC } from 'react';
import { useForm } from 'react-hook-form';
import { yupResolver } from '@hookform/resolvers/yup';
import * as yup from 'yup';

import Avatar from '@components/Avatar/Avatar';
import { StyledButton } from '@components/Button/styled';

import ImageIcon from '@icons/ImageIcon';

import { useTweetBlockHandlers } from '@hooks/useTweetBlockHandlers';

import {
  AvatarWrapper,
  ButtonWrapper,
  FileInput,
  ImageControlsWrapper,
  PreviewImage,
  TweetForm,
  TweetFooter,
  TweetInput,
  TweetWrapper,
} from './styled';

interface TweetBlockProps {
  setModal?: (value: boolean) => void;
}

export interface TweetFormType {
  tweet: string;
}

const schema = yup
  .object({
    tweet: yup.string().trim().min(3).required(),
  })
  .required();

const TweetBlock: FC<TweetBlockProps> = ({ setModal }) => {
  const { register, handleSubmit, reset } = useForm<TweetFormType>({
    resolver: yupResolver(schema),
  });

  const { previewUrl, inputRef, onSubmit, handleAddImage, handleChangeImage } =
    useTweetBlockHandlers(reset, setModal);

  return (
    <TweetWrapper>
      <AvatarWrapper>
        <Avatar />
      </AvatarWrapper>
      <TweetForm onSubmit={handleSubmit(onSubmit)}>
        <TweetInput {...register('tweet')} />
        <TweetFooter>
          <ImageControlsWrapper>
            <ImageIcon onClick={handleAddImage} />
            <FileInput ref={inputRef} onChange={handleChangeImage} />
            {previewUrl && <PreviewImage src={previewUrl} alt="Preview" />}
          </ImageControlsWrapper>
          <ButtonWrapper>
            <StyledButton $buttonType="primary">Tweet</StyledButton>
          </ButtonWrapper>
        </TweetFooter>
      </TweetForm>
    </TweetWrapper>
  );
};

export default TweetBlock;
