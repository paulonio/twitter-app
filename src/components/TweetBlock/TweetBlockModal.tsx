import React, { FC } from 'react';

import Modal from '@components/Modal/Modal';
import TweetBlock from '@components/TweetBlock/TweetBlock';

import { TweetBlockWrapper } from './styled';

interface TweetBlockModalProps {
  isActive: boolean;
  setActive: (value: boolean) => void;
}

const TweetBlockModal: FC<TweetBlockModalProps> = ({ isActive, setActive }) => {
  return (
    <Modal isActive={isActive} setActive={setActive}>
      <TweetBlockWrapper>
        <TweetBlock setModal={setActive} />
      </TweetBlockWrapper>
    </Modal>
  );
};

export default TweetBlockModal;
