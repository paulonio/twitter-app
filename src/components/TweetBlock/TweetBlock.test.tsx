import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

import TweetBlock from '@components/TweetBlock/TweetBlock';
import CustomProvider from '../EditUser/testUtils';

Object.defineProperty(URL, 'createObjectURL', {
  writable: true,
  value: jest.fn(),
});

describe('TweetBlock', () => {
  it('should render tweet block component', () => {
    const { container } = render(
      <CustomProvider>
        <TweetBlock />
      </CustomProvider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should change tweet input', () => {
    render(
      <CustomProvider>
        <TweetBlock />
      </CustomProvider>
    );

    const tweetInput: HTMLInputElement = screen.getByTestId('tweet-input');
    fireEvent.change(tweetInput, { target: { value: 'test' } });
    expect(tweetInput.value).toEqual('test');
  });

  it('should change image input', async () => {
    const file = new File(['hello'], 'hello.png', { type: 'image/png' });

    render(
      <CustomProvider>
        <TweetBlock />
      </CustomProvider>
    );

    const imageInput: HTMLInputElement = screen.getByTestId('image-input');
    await userEvent.upload(imageInput, file);

    expect(imageInput.files![0]).toStrictEqual(file);
    expect(imageInput.files!.item(0)).toStrictEqual(file);
    expect(imageInput.files).toHaveLength(1);
  });
});
