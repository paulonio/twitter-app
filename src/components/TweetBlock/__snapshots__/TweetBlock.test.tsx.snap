// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TweetBlock should render tweet block component 1`] = `
<div>
  <section
    class="sc-hBNyCF ketXfN"
  >
    <div
      class="sc-hlCNGy hNUIYK"
    >
      <img
        alt="Avatar"
        class="sc-erFrCC bFvuip"
        src="/assets/images/Avatar.svg"
      />
    </div>
    <form
      class="sc-hnHaRn lftyJx"
    >
      <input
        class="sc-QUirw btwCMJ"
        data-testid="tweet-input"
        name="tweet"
        placeholder="What's happening"
      />
      <div
        class="sc-SYvCl Ivzqc"
      >
        <div
          class="sc-gqHWMn hqEyGa"
        >
          <svg
            fill="currentColor"
            focusable="false"
            height="24"
            role="img"
            viewBox="0 0 24 24"
            width="24"
          >
            <path
              d="M19.75 2H4.25C3.01 2 2 3.01 2 4.25V19.75C2 20.99 3.01 22 4.25 22H19.75C20.99 22 22 20.99 22 19.75V4.25C22 3.01 20.99 2 19.75 2ZM4.25 3.5H19.75C20.163 3.5 20.5 3.837 20.5 4.25V13.926L16.642 10.068C16.502 9.928 16.312 9.848 16.112 9.848H16.109C15.909 9.848 15.716 9.928 15.577 10.072L11.26 14.456L9.44701 12.65C9.30701 12.51 9.11701 12.43 8.91701 12.43C8.72401 12.4 8.522 12.51 8.382 12.657L3.5 17.642V4.25C3.5 3.837 3.837 3.5 4.25 3.5ZM3.506 19.78L8.924 14.246L15.206 20.5H4.25C3.848 20.5 3.523 20.178 3.506 19.78ZM19.75 20.5H17.33L12.323 15.513L16.115 11.663L20.5 16.047V19.75C20.5 20.163 20.163 20.5 19.75 20.5Z"
              fill="#1D9BF0"
            />
            <path
              d="M8.86816 9.85097C9.72016 9.85097 10.4102 9.16067 10.4102 8.30897C10.4102 7.45737 9.72016 6.76697 8.86816 6.76697C8.01616 6.76697 7.32617 7.45737 7.32617 8.30897C7.32617 9.16067 8.01616 9.85097 8.86816 9.85097Z"
              fill="#1D9BF0"
            />
          </svg>
          <input
            accept="image/*"
            class="sc-cEPTvB QNYfJ"
            data-testid="image-input"
            type="file"
          />
        </div>
        <div
          class="sc-goDJBy kYDfmR"
        >
          <button
            class="sc-erzdQh iDymah"
            data-testid="tweet-button"
          >
            Tweet
          </button>
        </div>
      </div>
    </form>
  </section>
</div>
`;
